File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/DisplayItemDropUseCase.java
package com.habitrpg.android.habitica.interactors;

import android.os.Handler;
import androidx.appcompat.app.AppCompatActivity;
import android.view.ViewGroup;

import com.habitrpg.android.habitica.executors.PostExecutionThread;
import com.habitrpg.android.habitica.executors.ThreadExecutor;
import com.habitrpg.android.habitica.helpers.SoundManager;
import com.habitrpg.android.habitica.models.responses.TaskScoringResult;
import com.habitrpg.android.habitica.ui.views.HabiticaSnackbar;

import javax.inject.Inject;

import io.reactivex.Flowable;

public class DisplayItemDropUseCase extends UseCase<DisplayItemDropUseCase.RequestValues, Void> {

    private SoundManager soundManager;

    @Inject
    public DisplayItemDropUseCase(SoundManager soundManager, ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
        super(threadExecutor, postExecutionThread);
        this.soundManager = soundManager;
    }

    @Override
    protected Flowable<Void> buildUseCaseObservable(RequestValues requestValues) {
        return Flowable.defer(() -> {
            TaskScoringResult data = requestValues.data;

            if (data != null) {
                if (data.getDrop() != null) {
                    new Handler().postDelayed(() -> {
                        HabiticaSnackbar.Companion.showSnackbar(requestValues.snackbarTargetView,
                                data.getDrop().getDialog(), HabiticaSnackbar.SnackbarDisplayType.DROP);
                        soundManager.loadAndPlayAudio(SoundManager.SoundItemDrop);
                    }, 3000L);
                }
            }

            return Flowable.empty();
        });
    }

    public static final class RequestValues implements UseCase.RequestValues {

        private TaskScoringResult data;
        private AppCompatActivity context;
        private ViewGroup snackbarTargetView;

        public RequestValues(TaskScoringResult data, AppCompatActivity context, ViewGroup snackbarTargetView) {
            this.data = data;
            this.context = context;
            this.snackbarTargetView = snackbarTargetView;
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/interactors/DisplayItemDropUseCase.kt
package com.habitrpg.android.habitica.interactors

import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.helpers.SoundManager
import com.habitrpg.android.habitica.ui.views.HabiticaSnackbar
import com.habitrpg.shared.habitica.models.responses.TaskScoringResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject

class DisplayItemDropUseCase
    @Inject
    constructor(private val soundManager: SoundManager) :
    UseCase<DisplayItemDropUseCase.RequestValues, Unit>() {
        override suspend fun run(requestValues: RequestValues) {
            val data = requestValues.data
            val snackBarText = StringBuilder(data?.drop?.dialog ?: "")

            if ((data?.questItemsFound ?: 0) > 0 && requestValues.showQuestItems) {
                if (snackBarText.isNotEmpty()) {
                    snackBarText.append('\n')
                }
                snackBarText.append(
                    requestValues.context.getString(
                        R.string.quest_items_found,
                        data!!.questItemsFound,
                    ),
                )
            }

            if (snackBarText.isNotEmpty()) {
                MainScope().launch(context = Dispatchers.Main) {
                    delay(3000L)
                    HabiticaSnackbar.showSnackbar(
                        requestValues.snackBarTargetView,
                        snackBarText,
                        HabiticaSnackbar.SnackbarDisplayType.DROP,
                        true,
                    )
                    soundManager.loadAndPlayAudio(SoundManager.SOUND_ITEM_DROP)
                }
            }
            return
        }

        class RequestValues(
            val data: TaskScoringResult?,
            val context: AppCompatActivity,
            val snackBarTargetView: ViewGroup,
            val showQuestItems: Boolean,
        ) : UseCase.RequestValues
    }
